netket/optimizer/qgt/qgt_jacobian_pytree_logic.py

```diff
@@ -30,116 +31,133 @@
 
 
-def sub_mean(oks: PyTree) -> PyTree:
+# TODO better name and move it somewhere sensible
+def single_sample(forward_fn):
+    """
+    A decorator to make the forward_fn accept a single sample
+    """
+
+    def f(W, σ):
+        return forward_fn(W, σ[jnp.newaxis, :])[0]
+
+    return f
+
+
+# TODO move it somewhere reasonable
+def tree_subtract_mean(oks: PyTree) -> PyTree:
+    """
+    subtract the mean with MPI along axis 0 of every leaf
+    """
     return jax.tree_map(partial(subtract_mean, axis=0), oks)  # MPI
 
 
 @partial(jax.vmap, in_axes=(None, None, 0))
-def vmap_grad_real_holo(forward_fn: Callable, params: PyTree, samples: Array) -> PyTree:
+def jacobian_real_holo(forward_fn: Callable, params: PyTree, samples: Array) -> PyTree:
     """Calculates Jacobian entries by vmapping grad.
     Assumes the function is R→R or holomorphic C→C, so single grad is enough
 
     Args:
-        forward_fn: the log wavefunction, partialed with any model state, limited to single sample, taking a split real-imaginary PyTree in R→R mode
-        params : a pytree of parameters p, real-imaginary split in R→R mode
+        forward_fn: the log wavefunction ln Ψ
+        params : a pytree of parameters p
         samples : an array of n samples σ
 
     Returns:
         The Jacobian matrix ∂/∂pₖ ln Ψ(σⱼ) as a PyTree
     """
-    y, vjp_fun = jax.vjp(forward_fn, params, samples)
+    y, vjp_fun = jax.vjp(single_sample(forward_fn), params, samples)
     res, _ = vjp_fun(np.array(1.0, dtype=jnp.result_type(y)))
     return res
 
 
-def vmap_grad_centered_real_holo(
-    forward_fn: Callable, params: PyTree, samples: Array
-) -> PyTree:
-    """Calculates centred Jacobian (i.e., subtracts MPI mean from vmap_grad)"""
-    return sub_mean(vmap_grad_real_holo(forward_fn, params, samples))
-
-
 @partial(jax.vmap, in_axes=(None, None, 0))
-def vmap_grad_cplx(forward_fn: Callable, params: PyTree, samples: Array) -> PyTree:
+def jacobian_cplx(forward_fn: Callable, params: PyTree, samples: Array) -> PyTree:
     """Calculates Jacobian entries by vmapping grad.
     Assumes the function is R→C, backpropagates 1 and -1j
 
     Args:
-        forward_fn: the log wavefunction, partialed with any model state, limited to single sample, taking a split real-imaginary PyTree
-        params : a pytree of parameters p, real-imaginary split
+        forward_fn: the log wavefunction ln Ψ
+        params : a pytree of parameters p
         samples : an array of n samples σ
 
     Returns:
-        The Jacobian matrix ∂/∂pₖ ln Ψ(σⱼ) as a PyTree"""
-    y, vjp_fun = jax.vjp(forward_fn, params, samples)
+        The Jacobian matrix ∂/∂pₖ ln Ψ(σⱼ) as a PyTree
+    """
+    y, vjp_fun = jax.vjp(single_sample(forward_fn), params, samples)
     gr, _ = vjp_fun(np.array(1.0, dtype=jnp.result_type(y)))
     gi, _ = vjp_fun(np.array(-1.0j, dtype=jnp.result_type(y)))
-    return gr, gi
+    return jax.tree_multimap(jax.lax.complex, gr, gi)
```

Comment:
This wastes memory: you allocate enough memory to hold `gr_` and `gi_` (which will be quite big in production code!) when you
1. calculate them
2. turn them into a complex number
3. take the same complex numbers apart
4. stack them

I rewrote this part of the logic specifically to get rid of the two steps in the middle.

Reply:
In my mind when dreaming of the perfect jit compiler this should be optimized away (especially 2 and 3)...

I wanted to implement the real-imag split such that it only reinterprets the data at some point
https://github.com/netket/netket/blob/b8782770a5b1bb772c5ba6d4c0b607e814a51b50/netket/jax/utils.py#L240
which would avoid any copy in 2 and also 3 if done properly

and I also wanted to keep the flexibility of being able to change the implementation of `jacobian_cplx` later on where the most logical thing is that it returns the complex jacobian

However I guess we can do what you suggested for now with a few extra arguments here and there

Reply:
>In my mind when dreaming of the perfect jit compiler this should be optimized away (especially 2 and 3)...

I think that jits right now are not very smart about complex numbers.
However I like your idea! Maybe leave it in a comment?

Alas, from what I understood the best layout for complex numbers is SOA (a la Matlab) and not AOS like we/jax/C/everyone does, because you'd get this kind of things for free...